File: felicia/examples/learn/topic/camera/python/camera_subscribing_node.py
import felicia_py as fel
from felicia.core.protobuf.channel_pb2 import ChannelDef
from felicia.drivers.camera.camera_frame_message_pb2 import CameraFrameMessage


class CameraSubscribingNode(fel.NodeLifecycle):
    def __init__(self, camera_flag):
        super().__init__()
        self.camera_flag = camera_flag
        self.topic = camera_flag.topic_flag.value
        self.subscriber = fel.communication.Subscriber(CameraFrameMessage)

    def on_did_create(self, node_info):
        self.node_info = node_info
        self.request_subscribe()

    def on_message(self, message):
        print("CameraSubscribingNode.on_message()")

    def request_subscribe(self):
        settings = fel.communication.Settings()
        settings.period = fel.TimeDelta.from_seconds_d(
            1 / self.camera_flag.fps_flag.value)
        settings.is_dynamic_buffer = True

        self.subscriber.request_subscribe(self.node_info, self.topic,
                                          ChannelDef.CHANNEL_TYPE_TCP |
                                          ChannelDef.CHANNEL_TYPE_SHM,
                                          settings, self.on_message)

File: felicia/examples/learn/topic/camera/cc/camera_subscribing_node.cc
#include "felicia/examples/learn/topic/camera/cc/camera_subscribing_node.h"

namespace felicia {

CameraSubscribingNode::CameraSubscribingNode(const CameraFlag& camera_flag)
    : camera_flag_(camera_flag), topic_(camera_flag_.topic_flag()->value()) {}

void CameraSubscribingNode::OnDidCreate(NodeInfo node_info) {
  node_info_ = std::move(node_info);
  RequestSubscribe();
}

void CameraSubscribingNode::OnMessage(drivers::CameraFrameMessage&& message) {
  std::cout << "CameraSubscribingNode::OnMessage()" << std::endl;
}

void CameraSubscribingNode::RequestSubscribe() {
  communication::Settings settings;
  settings.queue_size = 1;
  settings.period =
      base::TimeDelta::FromSecondsD(1.0 / camera_flag_.fps_flag()->value());
  settings.is_dynamic_buffer = true;

  subscriber_.RequestSubscribe(
      node_info_, topic_,
      ChannelDef::CHANNEL_TYPE_TCP | ChannelDef::CHANNEL_TYPE_SHM, settings,
      base::BindRepeating(&CameraSubscribingNode::OnMessage,
                          base::Unretained(this)));
}

}  // namespace felicia